### PR TITLE
new output class and add configuration output in ".xyz" format

### PR DESCRIPTION
**Pull request by nw13slx**
A new class Output is set to replace the original output module. The class opens a log file and two xyz files at the init function and closes all files in conclude_run() function. All the original functions in the output module now become function members in the Output class, with the same arguments.
new class for output and write atomic configurations to xyz files
The new class avoid open/close the log file too often and also automatically back up the old log file if it exists before the current run starts.

write_xyz_config(...) function is also added to print configs in xyz format. It is called in the record_state function in otf.py and md_run.py.

Completely:
- issue #36

### CODE REVIEW

**Comment by nw13slx** [comment]
@jonpvandermause   all test ran well on my python environment on Odyssey... could you please show the syntax error? so I can figure out what's wrong?

**Comment by jonpvandermause** [comment]
Turns out I was 12 commits behind. This was the offending line (after commit 20374f9, since been corrected):
image

The tests work at the tip of the branch, but after running them the following files were generated and flagged by git after "git status":
image

Should we update the .gitignore so that these aren't tracked?

**Comment by nw13slx** [comment]
The latest commit should also solve issue #37 . All the hyper-parameter optimization output is redirected to a file with "-hyps.dat" suffix.

File: flare/qe_util.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 def run_espresso(qe_input, structure, pw_loc):
     run_qe_path = qe_input
     edit_qe_input_positions(run_qe_path, structure)
-    qe_command = 'mpirun {0} < {1} > {2}'.format(pw_loc, run_qe_path,
+    qe_command = '{0} < {1} > {2}'.format(pw_loc, run_qe_path,
```
Comment (stevetorr) [inline flare/qe_util.py]: This is a good catch, puts more power in the hand of the user to define the PWSCF command with/without MPI

File: flare/qe_util.py [hunk 2]
```diff
@@ -24,6 +24,8 @@ def run_espresso_par(qe_input, structure, pw_loc, no_cpus):
         'mpirun -np {0} {1} < {2} > {3}'.format(no_cpus, pw_loc, run_qe_path,
                                                 'pwscf.out')
 
+    with open("alog", "a+") as fout:
```
Comment (stevetorr) [inline flare/qe_util.py]: What does alog mean / stand for?

File: flare/gp_algebra.py [hunk 3]
```diff
@@ -278,20 +278,19 @@ def get_neg_likelihood(hyps: np.ndarray, training_data: list,
 
     like = get_like_from_ky_mat(ky_mat, training_labels_np)
 
-    if monitor:
-        print('like: ' + str(like))
-        print('\n')
+    if output is not None:
+        output.write_to_log('like: ' + str(like)+'\n', name="hyps")
```
Comment (stevetorr) [inline flare/gp_algebra.py]: A suggestion which is somewhat to personal preference: I personally find formatting strings using
` 'like: {} \n'.format(like) `
to be more readable.
I know that you also just translated the existing line this way, but going forward we could do all of our string formatting this way.

File: flare/gp_from_aimd.py [hunk 4]
```diff
@@ -244,11 +241,11 @@ def train_gp(self):
         """
         Train the Gaussian process and write the results to the output file.
         """
-        self.gp.train(monitor=True if self.verbose >= 2 else False)
+        self.gp.train(output=self.output if self.verbose >= 2 else None)
```
Comment (stevetorr) [inline flare/gp_from_aimd.py]: I like the way you implemented this.

File: flare/gp.py [hunk 5]
```diff
@@ -24,7 +24,7 @@ def __init__(self, kernel: Callable,
                  energy_kernel: Callable = None,
                  opt_algorithm: str = 'L-BFGS-B',
                  maxiter=10, par=False,
-                 monitor=False):
+                 output=None):
```
Comment (stevetorr) [inline flare/gp.py]: Here and in a few other places, we could typehint output as the `Output` object.

File: flare/output.py [hunk 6]
```diff
@@ -2,204 +2,320 @@
...
+        else:
+            self.outfiles[filetype] = open(filename, "w+")
+
+    def write_to_log(self, logstring: str, name: str = "log"):
+        """
+        Write any string to logfile
+        """
+        self.outfiles[name].write(logstring)
+
+    def write_header(self, cutoffs, kernel_name,
```
Comment (stevetorr) [inline flare/output.py]: In the future, we should consider writing a more general function that takes a dictionary in as an argument and then prints the key, value pairs in a similar way that the header method does (as well as a few other methods in this document which are tailored to individual use cases).

File: flare/gp.py (hunk 5 above)
Comment (jonpvandermause) [reply flare/gp.py]: Type hints will be eliminated eventually, as they don't play well with the Sphinx documentation. Might be better to do without them.

Comment (stevetorr) [reply flare/gp.py]: Sad to hear that Sphinx can't handle them. That's a shame. At least docstrings can contain the same info.

Comment (jonpvandermause) [reply flare/gp.py]: Actually it looks like Sphinx only really struggles with np.ndarray type hints. Here's an example from the docs:
image

If we can find a way to get Sphinx to format that better, type hints should be okay.

File: flare/qe_util.py (hunk 2 above)
Comment (nw13slx) [reply flare/qe_util.py]: ah... It's a line I left for debug. I forgot to delete it...

**Pull request merged**